ERROR 'sync port is not exposed in ws-machine. Workspace projects syncing is not possible'

Got bunch of these errors in log 
Need to include at least `workspaceID` in message
```
2017-02-24 09:01:16,460[aceSharedPool-5]  [ERROR] [c.c.machine.RemoteDockerNode 110]    - Sync port is not exposed in ws-machine. Workspace projects syncing is not possible
org.eclipse.che.api.core.ServerException: Sync port is not exposed in ws-machine. Workspace projects syncing is not possible
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.getSyncPort(DockerEnvironmentBackupManager.java:476) ~[codenvy-machine-hosted-5.3.1.jar:5.3.1]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspaceAndCleanup(DockerEnvironmentBackupManager.java:156) ~[codenvy-machine-hosted-5.3.1.jar:5.3.1]
        at com.codenvy.machine.RemoteDockerNode.unbindWorkspace(RemoteDockerNode.java:107) ~[codenvy-machine-hosted-5.3.1.jar:5.3.1]
        at org.eclipse.che.plugin.docker.machine.DockerInstance.destroy(DockerInstance.java:268) [che-plugin-docker-machine-5.3.1.jar:5.3.1]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine$MachineCleaner.lambda$onEvent$0(CheEnvironmentEngine.java:1398) [che-core-api-workspace-5.3.1.jar:5.3.1]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:28) ~[che-core-commons-lang-5.3.1.jar:5.3.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
```

**Reproduction Steps:**
Don't know, probably recipe for workspace doesn't contain `tail -f /dev/null` or equivalent command.

**Codenvy version:**    5.3.1

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/RemoteDockerNode.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/RemoteDockerNode.java
@@ -19,6 +19,7 @@ import com.codenvy.swarm.client.SwarmDockerConnector;
 import com.google.inject.assistedinject.Assisted;
 
 import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.environment.server.exception.EnvironmentException;
 import org.eclipse.che.api.machine.server.exception.MachineException;
 import org.eclipse.che.plugin.docker.client.DockerConnector;
 import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
@@ -90,7 +91,7 @@ public class RemoteDockerNode implements DockerNode {
     }
 
     @Override
-    public void bindWorkspace() throws ServerException {
+    public void bindWorkspace() throws ServerException, EnvironmentException {
         backupManager.restoreWorkspaceBackup(workspaceId, containerId, nodeHost);
         isBound = true;
     }
@@ -105,6 +106,8 @@ public class RemoteDockerNode implements DockerNode {
         } else {
             try {
                 backupManager.backupWorkspaceAndCleanup(workspaceId, containerId, nodeHost);
+            } catch (EnvironmentException e) {
+                LOG.info("Workspace unbinding failed due to environment error. Error: " + e.getLocalizedMessage());
             } catch (ServerException e) {
                 // TODO do throw it further when it won't brake ws stop
                 LOG.error(e.getLocalizedMessage(), e);

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/agent/launcher/MachineInnerRsyncAgentLauncherImpl.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/agent/launcher/MachineInnerRsyncAgentLauncherImpl.java
@@ -21,6 +21,7 @@ import org.eclipse.che.api.agent.server.launcher.CommandExistsAgentChecker;
 import org.eclipse.che.api.agent.server.launcher.CompositeAgentLaunchingChecker;
 import org.eclipse.che.api.agent.shared.model.Agent;
 import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.environment.server.exception.EnvironmentException;
 import org.eclipse.che.api.machine.server.spi.Instance;
 import org.eclipse.che.plugin.docker.machine.DockerInstance;
 import org.eclipse.che.plugin.docker.machine.node.DockerNode;
@@ -30,6 +31,7 @@ import javax.inject.Inject;
 import javax.inject.Named;
 import javax.inject.Singleton;
 
+import static java.lang.String.format;
 import static org.slf4j.LoggerFactory.getLogger;
 
 /**
@@ -66,7 +68,13 @@ public class MachineInnerRsyncAgentLauncherImpl extends AbstractAgentLauncher {
 
         DockerNode node = (DockerNode)machine.getNode();
         DockerInstance dockerMachine = (DockerInstance)machine;
-        node.bindWorkspace();
+        try {
+            node.bindWorkspace();
+        } catch (EnvironmentException e) {
+            throw new AgentStartException(
+                    format("Agent '%s' start failed because of an error with underlying environment. Error: %s",
+                           agent.getId(), e.getLocalizedMessage()));
+        }
         LOG.info("Docker machine has been deployed. " +
                  "ID '{}'. Workspace ID '{}'. " +
                  "Container ID '{}'. Node host '{}'. Node IP '{}'",

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/DockerEnvironmentBackupManager.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/DockerEnvironmentBackupManager.java
@@ -25,6 +25,7 @@ import org.eclipse.che.api.core.util.CommandLine;
 import org.eclipse.che.api.core.util.ListLineConsumer;
 import org.eclipse.che.api.core.util.ProcessUtil;
 import org.eclipse.che.api.core.util.ValueHolder;
+import org.eclipse.che.api.environment.server.exception.EnvironmentException;
 import org.eclipse.che.api.workspace.server.WorkspaceManager;
 import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
 import org.eclipse.che.plugin.docker.client.DockerConnector;
@@ -147,12 +148,14 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
      *         id of container that contains data
      * @param nodeHost
      *         host of a node where container is running
+     * @throws EnvironmentException
+     *         if backup failed due to abnormal state of machines in environment
      * @throws ServerException
-     *         if any error occurs
+     *         if any other error occurs
      */
     public void backupWorkspaceAndCleanup(String workspaceId,
                                           String containerId,
-                                          String nodeHost) throws ServerException {
+                                          String nodeHost) throws ServerException, EnvironmentException {
         try {
             String destPath = workspaceIdHashLocationFinder.calculateDirPath(backupsRootDir, workspaceId).toString();
             // if sync agent is not in machine port parameter is not used
@@ -185,12 +188,14 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
      *         id of container where data should be copied
      * @param nodeHost
      *         host of a node where container is running
+     * @throws EnvironmentException
+     *         if restore failed due to abnormal state of machines in environment
      * @throws ServerException
      *         if any error occurs
      */
     public void restoreWorkspaceBackup(String workspaceId,
                                        String containerId,
-                                       String nodeHost) throws ServerException {
+                                       String nodeHost) throws ServerException, EnvironmentException {
         try {
             String srcPath = workspaceIdHashLocationFinder.calculateDirPath(backupsRootDir, workspaceId).toString();
             User user = getUserInfo(workspaceId, containerId);
@@ -470,13 +475,16 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
      * @throws ServerException
      *         if port is not found
      */
-    private int getSyncPort(String containerId) throws ServerException {
+    private int getSyncPort(String containerId) throws ServerException, EnvironmentException {
         ContainerInfo containerInfo;
         try {
             containerInfo = dockerConnector.inspectContainer(containerId);
         } catch (IOException e) {
             LOG.error(e.getLocalizedMessage(), e);
             throw new ServerException("Workspace projects files in ws-machine are not accessible");
+        }
+        if (!containerInfo.getState().isRunning()) {
+            throw new EnvironmentException("Container " + containerId + " unexpectedly exited");
         }
 
         Map<String, List<PortBinding>> ports = firstNonNull(containerInfo.getNetworkSettings().getPorts(), emptyMap());

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/RemoteDockerNodeTest.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/RemoteDockerNodeTest.java
@@ -90,14 +90,14 @@ public class RemoteDockerNodeTest {
     }
 
     @Test
-    public void backupIsNotCalledWhenRestoreIsNotCalled() throws ServerException {
+    public void backupIsNotCalledWhenRestoreIsNotCalled() throws Exception {
         remoteDockerNode.unbindWorkspace();
 
         verifyBackupIsNeverCalled();
     }
 
     @Test
-    public void backupIsNotCalledWhenRestoreIsFailed() throws ServerException {
+    public void backupIsNotCalledWhenRestoreIsFailed() throws Exception {
         doThrow(new ServerException("no!"))
                 .when(backupManager)
                 .restoreWorkspaceBackup(anyString(), anyString(), anyString());
@@ -124,7 +124,7 @@ public class RemoteDockerNodeTest {
         remoteDockerNode.bindWorkspace();
     }
 
-    private void verifyBackupIsNeverCalled() throws ServerException {
+    private void verifyBackupIsNeverCalled() throws Exception {
         verify(backupManager, never()).backupWorkspaceAndCleanup(anyString(), anyString(), anyString());
     }
 }

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/backup/DockerEnvironmentBackupManagerTest.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/backup/DockerEnvironmentBackupManagerTest.java
@@ -17,6 +17,7 @@ package com.codenvy.machine.backup;
 import org.eclipse.che.api.core.NotFoundException;
 import org.eclipse.che.api.core.ServerException;
 import org.eclipse.che.api.core.model.machine.MachineStatus;
+import org.eclipse.che.api.environment.server.exception.EnvironmentException;
 import org.eclipse.che.api.machine.server.model.impl.MachineImpl;
 import org.eclipse.che.api.machine.server.model.impl.MachineRuntimeInfoImpl;
 import org.eclipse.che.api.machine.server.model.impl.ServerImpl;
@@ -28,6 +29,7 @@ import org.eclipse.che.plugin.docker.client.Exec;
 import org.eclipse.che.plugin.docker.client.LogMessage;
 import org.eclipse.che.plugin.docker.client.MessageProcessor;
 import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
+import org.eclipse.che.plugin.docker.client.json.ContainerState;
 import org.eclipse.che.plugin.docker.client.json.NetworkSettings;
 import org.eclipse.che.plugin.docker.client.json.PortBinding;
 import org.eclipse.che.plugin.docker.client.params.CreateExecParams;
@@ -210,6 +212,9 @@ public class DockerEnvironmentBackupManagerTest {
                                               singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
         ContainerInfo containerInfo = new ContainerInfo();
         containerInfo.setNetworkSettings(networkSettings);
+        ContainerState containerState = new ContainerState();
+        containerState.setRunning(true);
+        containerInfo.setState(containerState);
         when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
 
         executor = Executors.newFixedThreadPool(5);
@@ -304,9 +309,7 @@ public class DockerEnvironmentBackupManagerTest {
                 .thenReturn(singletonMap("23/tcp", new ServerImpl("ref", "proto", "127.0.0.1:" + PUBLISHED_SSH_PORT,
                                                                   null, null)));
 
-        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
-                                             CONTAINER_ID,
-                                             NODE_HOST);
+        backupManager.backupWorkspace(WORKSPACE_ID);
 
         // when
         backupManager.backupWorkspace(WORKSPACE_ID);
@@ -341,7 +344,7 @@ public class DockerEnvironmentBackupManagerTest {
         assertArrayEquals(RESTORE_WORKSPACE_COMMAND, command);
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldIgnoreNewBackupRequestIfPreviousOneHasBeenJustStarted() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -357,7 +360,7 @@ public class DockerEnvironmentBackupManagerTest {
         verify(backupManager, times(1)).executeCommand(anyObject(), anyInt(), anyString(), anyString());
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldIgnoreAllNewBackupRequestsIfOldHasNotFinishedYet() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -375,7 +378,8 @@ public class DockerEnvironmentBackupManagerTest {
 
     @Test(expectedExceptions = ServerException.class,
           expectedExceptionsMessageRegExp = "Restore of workspace " + WORKSPACE_ID +
-                                            " failed. Another restore process of the same workspace is in progress")
+                                            " failed. Another restore process of the same workspace is in progress",
+          timeOut = 2_000 * 60) // 2 minutes
     public void throwsExceptionOnNewRestoreIfAnotherOneIsInProgress() throws Exception {
         // given
         ThreadFreezer restoreFreezer = startNewProcessAndFreeze(this::runRestore);
@@ -400,7 +404,8 @@ public class DockerEnvironmentBackupManagerTest {
      */
     @Test(expectedExceptions = ServerException.class,
           expectedExceptionsMessageRegExp = "Restore of workspace " + WORKSPACE_ID +
-                                            " failed. Another restore process of the same workspace is in progress")
+                                            " failed. Another restore process of the same workspace is in progress",
+          timeOut = 2_000 * 60) // 2 minutes
     public void throwsExceptionOnNewRestoreIfAnotherOneIsInProgressAndAnotherRestoreFailed() throws Exception {
         // given
         ThreadFreezer restoreFreezer = startNewProcessAndFreeze(this::runRestore);
@@ -475,7 +480,7 @@ public class DockerEnvironmentBackupManagerTest {
                                              NODE_HOST);
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldThrowExceptionWhenStartRestoreIfBackupHasNotFinishedYet() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -507,6 +512,9 @@ public class DockerEnvironmentBackupManagerTest {
                                               singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
         ContainerInfo containerInfo = new ContainerInfo();
         containerInfo.setNetworkSettings(networkSettings);
+        ContainerState containerState = new ContainerState();
+        containerState.setRunning(true);
+        containerInfo.setState(containerState);
         when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
 
         // when
@@ -515,7 +523,27 @@ public class DockerEnvironmentBackupManagerTest {
                                              NODE_HOST);
     }
 
-    @Test
+    @Test(expectedExceptions = EnvironmentException.class,
+          expectedExceptionsMessageRegExp = "Container " + CONTAINER_ID + " unexpectedly exited")
+    public void restoreShouldThrowEnvironmentErrorIfContainerIsNoLongerRunning() throws Exception {
+        // given
+        NetworkSettings networkSettings = new NetworkSettings();
+        networkSettings.setPorts(singletonMap("22/tcp",
+                                              singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
+        ContainerInfo containerInfo = new ContainerInfo();
+        containerInfo.setNetworkSettings(networkSettings);
+        ContainerState containerState = new ContainerState();
+        containerState.setRunning(false);
+        containerInfo.setState(containerState);
+        when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
+
+        // when
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+    }
+
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldIgnoreOthersBackupsAndFailOnRestoresWhileBackupInProgress() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -540,7 +568,7 @@ public class DockerEnvironmentBackupManagerTest {
     }
 
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldIgnoreOthersBackupsAndFailOnRestoresWhileRestoreInProgress() throws Exception {
         // given
         ThreadFreezer restoreFreezer = startNewProcessAndFreeze(this::runRestore);
@@ -563,7 +591,7 @@ public class DockerEnvironmentBackupManagerTest {
         assertEquals(cmdCaptor.getValue()[0], RESTORE_SCRIPT);
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldIgnoreBackupIfRestoreHasNotFinishedYet() throws Exception {
         // given
         ThreadFreezer restoreFreezer = startNewProcessAndFreeze(this::runRestore);
@@ -579,7 +607,7 @@ public class DockerEnvironmentBackupManagerTest {
         assertEquals(cmdCaptor.getValue()[0], RESTORE_SCRIPT);
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldProcessOnlyOneBackupAtTheSameTime() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -605,7 +633,7 @@ public class DockerEnvironmentBackupManagerTest {
                                                 NODE_HOST);
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldQueuedBackupWithCleanupAfterBackup() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -647,7 +675,8 @@ public class DockerEnvironmentBackupManagerTest {
             backupManager.backupWorkspaceAndCleanup(WORKSPACE_ID,
                                                     CONTAINER_ID,
                                                     NODE_HOST);
-        } catch (RuntimeException ignored) {}
+        } catch (RuntimeException ignored) {
+        }
 
         // when
         backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
@@ -664,6 +693,9 @@ public class DockerEnvironmentBackupManagerTest {
                                               singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
         ContainerInfo containerInfo = new ContainerInfo();
         containerInfo.setNetworkSettings(networkSettings);
+        ContainerState containerState = new ContainerState();
+        containerState.setRunning(true);
+        containerInfo.setState(containerState);
         when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
 
         backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
@@ -676,7 +708,31 @@ public class DockerEnvironmentBackupManagerTest {
                                                 NODE_HOST);
     }
 
-    @Test
+    @Test(expectedExceptions = EnvironmentException.class,
+          expectedExceptionsMessageRegExp = "Container " + CONTAINER_ID + " unexpectedly exited")
+    public void cleanupShouldThrowEnvironmentErrorIfContainerIsNoLongerRunning() throws Exception {
+        // given
+        NetworkSettings networkSettings = new NetworkSettings();
+        networkSettings.setPorts(singletonMap("22/tcp",
+                                              singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
+        ContainerInfo containerInfo = new ContainerInfo();
+        containerInfo.setNetworkSettings(networkSettings);
+        ContainerState containerState = new ContainerState();
+        containerState.setRunning(false);
+        containerInfo.setState(containerState);
+        when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
+
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+
+        // when
+        backupManager.backupWorkspaceAndCleanup(WORKSPACE_ID,
+                                                CONTAINER_ID,
+                                                NODE_HOST);
+    }
+
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldBackupWithCleanupAfterFinishOfCurrentBackupButNotStartAnotherBackup()
             throws Exception {
         // given
@@ -697,7 +753,7 @@ public class DockerEnvironmentBackupManagerTest {
         assertEquals(cmdCaptor.getValue()[0], BACKUP_SCRIPT);
     }
 
-    @Test
+    @Test(timeOut = 2_000 * 60) // 2 minutes
     public void shouldIgnoreAllBackupsAndFailOnRestoresWhileBackupWithCleanupInQueueAfterBackup() throws Exception {
         // given
         injectWorkspaceLock(WORKSPACE_ID);
@@ -780,7 +836,7 @@ public class DockerEnvironmentBackupManagerTest {
             backupManager.backupWorkspaceAndCleanup(WORKSPACE_ID,
                                                     CONTAINER_ID,
                                                     NODE_HOST);
-        } catch (ServerException e) {
+        } catch (ServerException | EnvironmentException e) {
             LOG.error(e.getMessage());
         }
     }
@@ -790,7 +846,7 @@ public class DockerEnvironmentBackupManagerTest {
             backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
                                                  CONTAINER_ID,
                                                  NODE_HOST);
-        } catch (ServerException e) {
+        } catch (ServerException | EnvironmentException e) {
             LOG.error(e.getMessage());
         }
     }

